[OHFJIRA-6] : messages in final state

kkovarik opened the pull request:
## Final messages handling
**Motivation**: to have mechanism how to handle old messages in final state, that would otherwise be in datastore forever.

**Implemented solution**: job is deleting messages after some preconfigured time has elapsed from their last update time.

### Implementation
tried to keep it reasonably simple, yet customizable. There is job, service and collection of FinalMessageProcessors:

#### FinalMessagesProcessingJob
* job, configured to be executed only on one node of cluster
* does call FinalMessagesProcessingService

#### FinalMessagesProcessingService
* finds messages in db that are eligible for "processing" (aka delete)
* provided implementation does use configuration properties for each state
* states are hardcoded in current implementation, there is check that all parameters are configured
* does invoke all FinalMessageProcessors on each message, catches exceptions so that failure to delete one message should not affect the others
* all the processors are invoked in single transaction

#### interface FinalMessageProcessor
* all beans of this type will be applied to message
* order is defined expicitely & respected

#### DeleteFinalMessageProcessor implements FinalMessageProcessor
* processor, enabled/disabled by property
* does delete provided message with all the related entities - extcalls & requests & responses.
 
#### AbstractFinalMessageProcessor
* abstract class to ease FinalMessageProcessor implementation

#### Notes
* had to add some delete methods to Dao´s, did not added them to Service layer, as they are only called from DeleteMessageProcessor
* job and all the beans are conditionaly registered by using property `ohf.asynch.finalMessages.processingEnabled`, by default disabled
* minor backward compatible change in MessageFilter, to enable search for messages by state & lastUpdate only, not to have to add receivedFrom

### Configuration
```
# final messages processing enabled or disabled. Note: cannot be defined in database, 
# needs to be in properties.
ohf.asynch.finalMessages.processingEnabled = true/false

# final messages processing job interval, in seconds.
ohf.asynch.finalMessages.processingIntervalSec=

# maximum number of messages processed per job.
ohf.asynch.finalMessages.iterationMessageLimit=10000

# duration in seconds to keep messages in OK state in the datastore. After this period expires and
# there is no other action with the message, it will be processed as final (deleted probably)
ohf.asynch.finalMessages.ok.saveTimeInSec=0

# duration in seconds to keep messages in CANCEL state in the datastore.
ohf.asynch.finalMessages.cancel.saveTimeInSec=28800

# duration in seconds to keep messages in FAILED state in the datastore.
ohf.asynch.finalMessages.failed.saveTimeInSec=14400

# enable/disable processor that does delete messages
ofh.asynch.finalMessages.deleteProcessor.enabled=true
```
### Possible pittfals 
* performance - how many messages should be processed in single iteration? when enabled on already full database? now there is configurable limit. however it is fetched by single query, could make sense to make multiple queries if this parameter should be too big...

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessagesProcessingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+/*
```
add package-info.java files into packages ...

kkovarik replied:
Thanks, added

pjuza commented on core-api/src/main/java/org/openhubframework/openhub/api/configuration/CoreProps.java:
```diff
@@ -82,6 +82,37 @@
+     * Duration in seconds to keep messages in OK state in the datastore. After this period expires and
+     * there is no other action with the message, it will be processed as final (deleted probably).
+     */
+    public static final String ASYNCH_FINAL_MESSAGES_OK_SAVE_TIME_IN_SEC = PREFIX + "asynch.finalMessages.ok.saveTimeInSec";
```
What does it mean if value is 0? Please describe it.
Maybe this comment is mainly for description on wiki to be clear what does it mean to use 0 value for time intervals.

kkovarik replied:
Setting to '0' means, that messages will be processed as soon as possible. Effectively meaning next scheduled job run,

Javadoc improved, thanks good idea to pointing that out.

pjuza replied:
And what value should I use if I want to have specific messages forever in data store?

kkovarik replied:
Added possibility to set value to "-1", then it will be kept indefinitely

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessagesProcessingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+
+    protected static final Map<MsgStateEnum, String> FINAL_MSG_STATES_CONFIG = new HashMap<>();
+    static {
+        FINAL_MSG_STATES_CONFIG.put(MsgStateEnum.OK, CoreProps.ASYNCH_FINAL_MESSAGES_OK_SAVE_TIME_IN_SEC);
```
It would be nice to have definition of final states at one place only. See org.openhubframework.openhub.api.entity.MsgStateEnum#isFinal

kkovarik replied:
Ok, if you think it will ever by changed It´s small change as I was anticipating it :) 

I have to relative the relationship between State + Prop, so I will have to make dynamic resolving like:
`asynch.finalMessages.<lowercased-name-of-state>.saveTimeInSec`, ok?

What should happen if configuration value is not found, exception during startup? Other solution would be default value, but I´m not too fond of it.

pjuza replied:
I would throw exception during startup ...

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessagesProcessingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @ConfigurableValue(key = CoreProps.ASYNCH_FINAL_MESSAGES_ITERATION_MESSAGE_LIMIT)
+    protected ConfigurationItem<Long> messagesTechnicalLimit;
```
I like using limit, good idea

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/DeleteFinalMessageProcessor.java:
```diff
@@ -0,0 +1,158 @@
+    @Override
+    public void processMessage(final Message message) {
+        // notify listeners, note they are not wrapped in try catch,
+        // so if listener fails, it will effectivelly cancel the delete
```
Mispelling - effectively

kkovarik replied:
Fixed, thanks

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/DeleteFinalMessageProcessor.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Override
+    public void processMessage(final Message message) {
```
I had different idea how to implemented it - I wanted to define interface, similar to FinalMessageProcessor in your implementation. And this interface will have default implementation that deletes records from data store. And if you would like to change it for your project then you will implement it in your way. 
But your implementation deletes records in every case, only custom implementations before deletion itself are possible.

hanusto replied:
But we have to force to delete outdated messages. If this listener will be Spring component, it will be possible by Primary annotation but it should be exceptionally.

kkovarik replied:
@pjuza  thanks for mentioning it, DeleteFinalMessageProcessor is now ConditionalOnMissing bean, meaning it can be overriden easily, not just with Primary bean. Then you can simple achieve different behaviour with your custom FinalMessageProcessor.

However I think in most cases, custom implementation will want to delete the messages ultimately. Just implement custom "archiving" before that. That is achievable with the listener only. Listener has some control as well, meaning, if it throws exception, delete will not continue. That way it could hopefully ensure, that messages won´t be lost unexpectedly.

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/DeleteFinalMessageProcessor.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Override
+    public void processMessage(final Message message) {
```
I would add final modifier to this method - this method represents "core" of this class and should be unchangeble. You can override each delete method but core logic should be always same.

kkovarik replied:
Thanks, good point. If this is overriden, the very "core" will be changed.

Changed.

hanusto commented on core-api/src/main/java/org/openhubframework/openhub/api/asynch/finalmessage/FinalMessageProcessor.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Contract for final messages processor.
+ * Messages that are no-longer considered relevant for openhub async workflow are expected
+ * to be processed by this processor. After this, they should not be in OpenHub datastore
```
According to comment from @pjuza it is wrong javadoc. _FinalMessageProcess_ is only about processing messages in final state. Removing or archiving is one of many action what can be done.

kkovarik replied:
Thanks, good point, updated javadoc.

hanusto commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/DeleteFinalMessageProcessor.java:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @return list of listeners, that will be executed consequently.
+     */
+    protected List<DeleteFinalMessageListener> getSortedListenerList() {
```
Why? If each component has Order annotation, it is sorted by default.

hanusto replied:
I see, this is safer. I agree.

hanusto commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/DeleteFinalMessageProcessor.java:
```diff
@@ -0,0 +1,158 @@
+    protected List<DeleteFinalMessageListener> sortedListenerList;
+
+    @Autowired
+    public DeleteFinalMessageProcessor(PlatformTransactionManager transactionManager) {
```
Which _ PlatformTransactionManager_? OpenHub?

kkovarik replied:
Well, good question, was thinking about it as well. Relates to the PR https://github.com/OpenWiseSolutions/openhub-framework/pull/93 

It should be the transactionManager that manages openhub datasource. Which one is that? In proposition of https://github.com/OpenWiseSolutions/openhub-framework/pull/93 it is either Spring-Boot autoconfigured one (then only one is present), or the Primary bean, if multiple beans are present.

Hence wiring with the interface only, not with qualifier or anything.

hanusto commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/DeleteFinalMessageProcessor.java:
```diff
@@ -0,0 +1,158 @@
+ * @see FinalMessageProcessor
+ * @see DeleteFinalMessageListener
+ */
+public class DeleteFinalMessageProcessor implements FinalMessageProcessor {
```
It is not spring component?

kkovarik replied:
It is, however it is registered in Configuration class - see FinalMessageProcessingConfiguration, motivation was to register it only if this whole functionality is enabled

hanusto commented on core/src/main/resources/db/migration/postgresql/V1_0_3__final_messages.sql:
```diff
@@ -0,0 +1,27 @@
+-- core.async
+--
+
+-- final messages processing enabled or disabled.
```
In future we should migrate from Flyway to Liquibase...

hanusto commented on core/src/test/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessagesProcessingServiceImplTest.java:
```diff
@@ -0,0 +1,208 @@
+
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.openhubframework.openhub.api.asynch.finalmessage.FinalMessageProcessor;
```
In future we can reorder imports = it will be necessary to improve our code style.

kkovarik committed:
[OHFJIRA-6] : processing of messages in final states (OK, CANCEL, FAILED)

- added job to search for messages
- added implementation of processor to delete messages from db
- new config params to control this feature added

kkovarik commented:
Updated implementation & description. Now there is job, service and collection of FinalMessageProcessors. One of the MessageProcessor is DeleteFinalMessageProcessor. Please check PR again.

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessagesProcessingJob.java:
```diff
@@ -0,0 +1,64 @@
+                    repeatIntervalProperty = CoreProps.ASYNCH_FINAL_MESSAGES_PROCESSING_INTERVAL_SEC,
+                    intervalPropertyUnit = SimpleTriggerPropertyUnit.SECONDS)
+    )
+    public final void processFinalMessages() {
```
I would appreacite if I could disable job at all. Maybe also with some configuration parameter ...
I would use the same parameter you already have: org.openhubframework.openhub.api.configuration.CoreProps#ASYNCH_FINAL_MESSAGES_PROCESSING_ENABLED

kkovarik replied:
It is already implemented that way. Job is not on packageScan, it is registered as bean conditionaly in the configuration class **org.openhubframework.openhub.core.common.asynch.finalmessage.FinalMessageProcessingConfiguration**

idea is that none if this beans (job, service, processor) is registered, if the property  ASYNCH_FINAL_MESSAGES_PROCESSING_ENABLED is false

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessagesProcessingServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+                } catch (Exception ex) {
+                    // continue with another message if handling of one fails.
+                    LOG.warn("Failed to process message : {}, will continue with next.", message, ex);
```
I think it's "normal" error => please change log level to error.

kkovarik replied:
Changed

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/AbstractFinalMessageProcessor.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    @Override
+    public void processMessage(final Message message) {
```
this method should be final

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/AbstractFinalMessageProcessor.java:
```diff
@@ -0,0 +1,62 @@
+    public void processMessage(final Message message) {
+        // fetch the message to have it in current hibernate session
+        final Message loaded = messageDao.findMessage(message.getId());
+
```
I would add logging on level DEBUG/TRACE ...

kkovarik replied:
Thanks for mentioning it, added logging to AbstractFinalMessageProcessor and added seting MDC Log context in FinalMessageProcessingService, so that logs of message processing are correlated via correlationId.

pjuza commented on core/src/main/java/org/openhubframework/openhub/core/common/asynch/finalmessage/FinalMessageProcessingConfiguration.java:
```diff
@@ -0,0 +1,58 @@
+import org.openhubframework.openhub.api.asynch.finalmessage.FinalMessageProcessor;
+import org.openhubframework.openhub.api.asynch.finalmessage.FinalMessagesProcessingService;
+import org.openhubframework.openhub.api.configuration.CoreProps;
+import org.springframework.beans.factory.annotation.Qualifier;
```
Unused imports ...

kkovarik replied:
Thanks!

pjuza commented:
Please dont't forget to add new wiki page - I prefer to add new page under 
https://openhubframework.atlassian.net/wiki/spaces/OHF/pages/33573/User+guide. It's quite new feature :)

kkovarik commented:
Added documentation https://openhubframework.atlassian.net/wiki/spaces/OHF/pages/778764295/Messages+in+final+state and configuration described in https://openhubframework.atlassian.net/wiki/spaces/OHF/pages/33599/OpenHub+configuration

pjuza commented:
My last two comments to AbstractFinalMessageProcessor - I would change logging level to trace from debug because often there is huge amount of messages and I would add add final to processMessage method.
I like your solution!

kkovarik commented:
Log level changed & processMessage not final
